src/LimitedLDLFactorizations.jl

```diff
@@ -51,70 +54,38 @@ mutable struct LimitedLDLFactorization{
   indr::Vector{Ti}
   indf::Vector{Ti}
   list::Vector{Ti}
-  pos::Vector{Ti}
-  neg::Vector{Ti}
+  pos::Vector{Int}
+  neg::Vector{Int}
 
   computed_posneg::Bool # true if pos and neg are computed (becomes false after factorization)
 end
 
-lldl(A::Array{Tv, 2}; kwargs...) where {Tv <: Number} = lldl(sparse(A); kwargs...)
-
-"""
-    lldl(A)
-
-Compute the limited-memory LDLᵀ factorization of A without pivoting.
-
-# Arguments
-- `A::SparseMatrixCSC{Tv,Ti}`: matrix to factorize (its strict lower triangle and
-                               diagonal will be extracted)
-
-# Keyword arguments
-- `memory::Int=0`: extra amount of memory to allocate for the incomplete factor `L`.
-                   The total memory allocated is nnz(T) + n * `memory`, where
-                   `T` is the strict lower triangle of A and `n` is the size of `A`.
-- `α::Tv=Tv(0)`: initial value of the shift in case the incomplete LDLᵀ
-                 factorization of `A` is found to not exist. The shift will be
-                 gradually increased from this initial value until success.
-- `droptol::Tv=Tv(0)`: to further sparsify `L`, all elements with magnitude smaller
-                       than `droptol` are dropped.
-"""
-function lldl(A::SparseMatrixCSC{Tv, Ti}; kwargs...) where {Tv <: Number, Ti <: Integer}
-  lldl(tril(A, -1), diag(A), amd(A); kwargs...)
-end
-
-function lldl(
-  A::SparseMatrixCSC{Tv, Ti},
-  P::AbstractVector{<:Integer};
-  kwargs...,
-) where {Tv <: Number, Ti <: Integer}
-  lldl(tril(A, -1), diag(A), P; kwargs...)
-end
-
-# symmetric matrix input
-function lldl(
-  sA::Symmetric{T, SparseMatrixCSC{T, Ti}},
-  args...;
-  kwargs...,
-) where {T <: Real, Ti <: Integer}
-  sA.uplo == 'U' && error("matrix must contain the lower triangle")
-  A = sA.data
-  lldl(A, args...; kwargs...)
-end
-
 function LimitedLDLFactorization(
-  adiag::AbstractVector{Tv},
+  T::SparseMatrixCSC{Tv, Ti},
   P::AbstractVector{<:Integer},
-  Ti::DataType,
   memory::Int,
   α::Tv,
   n::Int,
   nnzT::Int,
-) where {Tv <: Number}
+) where {Tv <: Number, Ti}
   np = n * memory
   Pinv = similar(P)
 
+  nb_diag_elements = 0
+  adiag = Vector{Tv}(undef, n)
+  for col = 1:n
+    k = T.colptr[col]
+    row = (k ≤ nnzT) ? T.rowval[k] : 0
+    if row == col
+      nb_diag_elements += 1
+      adiag[col] = T.nzval[k]
+    else
+      adiag[col] = zero(Tv)
+    end
+  end
```

Comment:
```suggestion
  adiag = Vector{Tv}(undef, n)
  for col = 1:n
    find = false
    ind = T.colptr[col]
    while !find && ind ≤ T.colptr[col+1]-1
      row = T.rowval[ind]
      if row == col
        nnz_diag += 1
        adiag[col] = T.nzval[ind]
      else
        ind += 1
      end
    end
    if !find
      adiag[col] = zero(Tv)
    end
  end
```

Reply:
I didn't understand `row = (k ≤ nnzT) ? T.rowval[k] : 0`.
What is the goal of this line?

Reply:
As T is lower triangular, the diagonal is the first element of each column: `T.colptr[col]`. However, there could be an implicit zero on the column `col`. That's why I check `row == col` (i.e. `T.rowval[row] == T.colptr[col]` with `row = T.colptr[col]`).
The line `row = (k ≤ nnzT) ? T.rowval[k] : 0` is useful in case at least one of the last columns of `T` is full of zeros. For any of these last columns `col`, we have `T.colptr[col] = T.colptr[end] = nnzT + 1`. In this case `T.rowval[col]` is not accessible, so I set `row = 0` to ensure that `adiag[col] = 0`.
I can add a comment if this is not clear.

Reply:
I forgot that we must do `tril(A)` in this package and we don't have any non-zero in the upper triangle.
It will be great to add a comment that explain why `T.rowval[k]` is the diagonal coefficient if it's a non-zero.

Reply:
You could just add:
`As T is lower triangular, the diagonal is the first element of each column: T.colptr[col]`